src/mozanalysis/experiment.py

```diff
@@ -309,6 +309,33 @@ def get_per_client_data(
             data_source, time_limits
         ).alias('data_source')
 
+        join_on = self._get_join_conditions(enrollments, data_source, time_limits)
+
+        sanity_metrics = self._get_telemetry_sanity_check_metrics(
+            enrollments, data_source
+        )
+
+        res = enrollments.join(
+            data_source,
+            join_on,
+            'left'
+        ).groupBy(
+            enrollments.client_id, enrollments.branch
+        ).agg(
+            *(metric_list + sanity_metrics)
+        )
+        if keep_client_id:
+            return res
+        else:
+            return res.drop(enrollments.client_id)
+
+    def _get_join_conditions(self, enrollments, data_source, time_limits):
+        """Return a list of join conditions.
+
+        In ``_get_results_for_one_data_source``, we join ``enrollments``
+        to ``data_source``. This method returns the list of join
+        conditions.
```

Comment:
docs nit: returns a list of boolean `Column`s representing join conditions between the enrollments `DataFrame` and data_source. When used with `DataFrame.join` to perform a left-join of the data_source onto the enrollments, these conditions produce a `DataFrame` containing the rows from data_source for enrolled users, from the time of their enrollment until the end of the experiment.

Reply:
"from the time of their enrollment until the end of the experiment" is not quite correct in general - it only contains the rows received during the analysis window, which is typically a subset of that time. But I've incorporated the gist of your message:
```lang=py
        """Return a list of join conditions.

        Returns a list of boolean ``Column``s representing join
        conditions between the ``enrollments`` ``DataFrame`` and
        ``data_source``.

        In ``_get_results_for_one_data_source``, we left join
        ``enrollments`` to ``data_source`` using these join conditions
        to produce a ``DataFrame`` containing the rows from
        ``data_source`` for enrolled clients that were submitted during
        the analysis window.
        """
```